packages/trader/src/App/Components/Elements/market-countdown-timer.jsx

```diff
@@ -0,0 +1,212 @@
+            symbol = symbols.find(item => item.symbol === target_symbol);
+            if (symbol !== undefined) return symbol;
+        }
+    }
+};
+
+const calculateTimeLeft = remaining_time_to_open => {
+    const difference = remaining_time_to_open - +new Date();
+    return difference > 0
+        ? {
+              days: Math.floor(difference / (1000 * 60 * 60 * 24)),
+              hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
+              minutes: Math.floor((difference / 1000 / 60) % 60),
+          }
+        : {};
+};
+
+const MarketCountdownTimer = ({
+    is_main_page,
+    is_market_close_overlay_loading,
+    setActiveSymbols,
+    setMarketCloseOverlayLoading,
+    symbol,
+}) => {
+    const isMounted = useIsMounted();
+    const [when_market_opens, setWhenMarketOpens] = React.useState({});
+    const [time_left, setTimeLeft] = React.useState(calculateTimeLeft(when_market_opens?.remaining_time_to_open));
+
+    React.useEffect(() => {
+        // eslint-disable-next-line consistent-return
+        const whenMarketOpens = async (days_offset, target_symbol) => {
+            // days_offset is 0 for today, 1 for tomorrow, etc.
+            if (days_offset > days_to_check_before_exit) return {};
+            let remaining_time_to_open;
+            const target_date = moment(new Date()).add(days_offset, 'days');
+            const api_res = await getTradingTimes(target_date.format('YYYY-MM-DD'));
+            if (!api_res.api_initial_load_error) {
+                const { times } = getSymbol(target_symbol, api_res.trading_times);
+                const { open, close } = times;
+                const is_closed_all_day = open?.length === 1 && open[0] === '--' && close[0] === '--';
+                if (is_closed_all_day) {
+                    // check tomorrow trading times
+                    return whenMarketOpens(days_offset + 1, target_symbol);
+                }
+                const date_str = target_date.toISOString().substring(0, 11);
+                const getUTCDate = hour => new Date(`${date_str}${hour}Z`);
+                for (let i = 0; i < open?.length; i++) {
+                    const diff = +getUTCDate(open[i]) - +new Date();
+                    if (diff > 0) {
+                        remaining_time_to_open = +getUTCDate(open[i]);
+                        if (isMounted() && target_symbol === symbol) {
+                            return setWhenMarketOpens({ days_offset, opening_time: open[i], remaining_time_to_open });
+                        }
+                    }
+                }
+                whenMarketOpens(days_offset + 1, target_symbol);
+            }
+        };
+
+        whenMarketOpens(0, symbol);
+    }, [isMounted, symbol]);
+
+    React.useEffect(() => {
+        let timer;
+        if (when_market_opens?.remaining_time_to_open) {
+            timer = setTimeout(async () => {
+                setTimeLeft(calculateTimeLeft(when_market_opens?.remaining_time_to_open));
```

Comment:
- You may want to wrap this in an `isMounted`check to avoid updating state on an unmounted component since this is ran asynchronously.
- You may not need the optional chaining for `remaining_time_to_open` as you've already checked whether it has a value on line 93, unless there's a chance the value changes.